services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/controller/SecurityGroupRuleController.java

```diff
@@ -0,0 +1,250 @@
+            throw new SecurityGroupRuleIdRequired();
+        }
+    }
+
+    private void checkPortRange(SecurityGroupRule securityGroupRule) throws Exception {
+        Integer portRangeMax = securityGroupRule.getPortRangeMax();
+        Integer portRangeMin = securityGroupRule.getPortRangeMin();
+        String protocol = securityGroupRule.getProtocol();
+
+        if (protocol == null) {
+            throw new ProtocolInvalidException();
+        }
+
+        if (Arrays.asList("tcp", "udp").contains(protocol)) {
+            if (portRangeMax == null || portRangeMin == null) {
+                throw new PortRangeInvalid();
+            }
+
+            if (!(portRangeMax > 0 && portRangeMax < 65536)) {
+                throw new PortRangeInvalid();
+            }
+
+            if (!(portRangeMin > 0 && portRangeMin < 65536)) {
+                throw new PortRangeInvalid();
+            }
+
+            if (portRangeMin > portRangeMax) {
+                throw new PortRangeInvalid();
+            }
+        }
+
+        if (Arrays.asList("icmp", "icmpv6").contains(protocol)) {
+            if (portRangeMin == null && portRangeMax != null) {
+                throw new PortRangeInvalid();
+            }
+
+            if (portRangeMin != null && portRangeMax == null) {
+                throw new PortRangeInvalid();
+            }
+
+            if (portRangeMin != null && !(0 <= portRangeMin && portRangeMin <= 255)) {
+                throw new PortRangeInvalid();
+            }
+
+            if (portRangeMax != null && !(0 <= portRangeMax && portRangeMax <= 255)) {
+                throw new PortRangeInvalid();
+            }
+        }
+    }
+
+    private String getIpVersionByPrefix(String prefix) throws Exception {
+        if (Ipv4AddrUtil.ipv4PrefixCheck(prefix)) {
+            return "IPv4";
+        } else if (Ipv6AddrUtil.ipv6PrefixCheck(prefix)) {
+            return "IPv6";
+        }
+
+        throw new RemoteIpPrefixInvalid();
+    }
+    private void checkRemoteIpPrefix(SecurityGroupRule securityGroupRule) throws Exception {
+        String remoteIpPrefix = securityGroupRule.getRemoteIpPrefix();
+        String etherType = securityGroupRule.getEtherType();
+
+        if (!getIpVersionByPrefix(remoteIpPrefix).equals(etherType)) {
+            throw new EtherTypeRemoteIpPrefixConflict();
+        }
+    }
+
+    private void checkProtocolAndEtherType(SecurityGroupRule securityGroupRule) throws Exception {
+        String protocol = securityGroupRule.getProtocol();
+        String etherType = securityGroupRule.getEtherType();
+
+        List<String> ipv6Protocols = Arrays.asList("icmpv6");
+        if (ipv6Protocols.contains(protocol) &&
+                SecurityGroupRule.EtherType.IPV4.getType().equals(etherType)) {
+            throw new ProtocolEtherTypeConflict();
+        }
+    }
+
+    private void checkSecurityGroupRule(SecurityGroupRule securityGroupRule) throws Exception {
```

Comment:
Didn't see protocol check here. Which protocols are supported by the v1 implementation. Could you list the current scope at the beginning of the controller.

Reply:
Oh, let check.